src/Type/IntersectionType.php

```diff
@@ -799,6 +799,9 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $uni
 		}
 
 		$result = $this->intersectTypes(static fn (Type $type): Type => $type->setOffsetValueType($offsetType, $valueType, $unionValues));
+		if ($this->isList()->yes() && $valueType->isArray()->yes()) {
+			$result = TypeCombinator::intersect($result, new AccessoryArrayListType());
```

Comment:
Again, I don't get why does it matter what `$valueType` is.

Reply:
the idea was:

```
/** @param list<array<int>> $list */
function doFoo($list, int $i) {
  $list[$i][3] = 123;
}
```

when setting offset $i on $list to `array(3 => 123)`, it does not turn $list into array.

Reply:
But that's true for anything, not just arrays. I feel like we should call `setExistingOffsetValueType` more if we're sure the offset exists.

Reply:
This just doesn't make sense. A simple code:

```php
/** @param list<array<int>> $list */
function doFoo($list, int $i) {
	$list[$i][3] = 123;
	\PHPStan\dumpType($list);
}
```

We don't know whether `$list[$i]` exists. 2.1.x is correct to dump this as `non-empty-array<int, non-empty-array<int>&hasOffsetValue(3, 123)>`.

Your branch dumps this as `non-empty-list<non-empty-array<int>&hasOffsetValue(3, 123)>` which is wrong.

Reply:
For example, with code like this:

```php
foreach ($list as $i => $value) {
    $list[$i][3] = 123;
}
```

We should preserve `$list` being a list because `$list[$i]` exists and this information is already carried in Scope (hasExpressionType).

Right now `setExistingOffsetValueType` is only called for `unset()` but we could call it for when we're sure the offset exists.

Reply:
I am now using `setExistingOffsetValueType` more, but in the process somehow broke some tests.
need to investigate further, but as I kind of working thru the fog was thinking it would be worth sharing the progress to verify I am working in the right direction